### PR TITLE
Return HTTP 404 when FileNotFoundException is thrown when attempting to send a file

### PR DESCRIPTION
**Pull request by natemcmaster**
Resolves https://github.com/aspnet/Home/issues/2775

If a file disappears between the File.Exists and File.Open, or if the file is a broken symlink, static files throws an uncaught exception, causing HTTP 500. This change catches the FileNotFoundExcpetion and returns HTTP 404 instead.

### CODE REVIEW

File: src/Microsoft.AspNetCore.StaticFiles/StaticFileMiddleware.cs [hunk 1]
```diff
@@ -97,21 +98,35 @@ public Task Invoke(HttpContext context)
...
+                        }
+                        try
+                        {
+                            await fileContext.SendAsync();
+                            _logger.LogFileServed(fileContext.SubPath, fileContext.PhysicalPath);
+                            return;
+                        }
+                        catch (FileNotFoundException)
+                        {
+                            _logger.LogFileNotFound(fileContext.SubPath);
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.StaticFiles/StaticFileMiddleware.cs]: This does not reset the headers set by SendAsync. Either reset the headers at this point or delay setting headers in SendAsync

Comment (natemcmaster) [reply src/Microsoft.AspNetCore.StaticFiles/StaticFileMiddleware.cs]: Good catch. I added test coverage for this and delayed setting headers until after the FileNotFoundException would have been thrown

File: src/Microsoft.AspNetCore.StaticFiles/StaticFileMiddleware.cs [hunk 2]
```diff
@@ -97,21 +98,35 @@ public Task Invoke(HttpContext context)
...
+                            await fileContext.SendStatusAsync(StatusCodes.Status404NotFound);
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.StaticFiles/StaticFileMiddleware.cs]: Returning an explicit 404 is a strange choice. There's no other condition where this middleware returns 404. Calling next would be better.

File: src/Microsoft.AspNetCore.StaticFiles/StaticFileMiddleware.cs (hunk 1 above)
Comment (Tratcher) [reply src/Microsoft.AspNetCore.StaticFiles/StaticFileMiddleware.cs]: You didn't do this for the SendFileAsync code path, nor is it possible to delay setting the headers.

**Comment by Tratcher** [comment]
@natemcmaster this is incomplete/buggy. Revert and we'll keep iterating.

**Comment by natemcmaster** [comment]
https://github.com/aspnet/StaticFiles/commit/3cd6add095e8a4feccbd7f0886ab3a2c293cfd5f

**Pull request merged**